File: src/test/java/com/openvalue/boardgameratings/service/rating/BoardGameRatingServiceTest.java
package com.openvalue.boardgameratings.service.rating;

import com.openvalue.boardgameratings.api.BoardGame;
import com.openvalue.boardgameratings.api.Rating;
import com.openvalue.boardgameratings.api.request.RatingRequest;
import com.openvalue.boardgameratings.service.boardgame.BoardGameRepository;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import static com.openvalue.boardgameratings.service.util.TestData.*;
import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class BoardGameRatingServiceTest {

    @Mock
    private BoardGameRepository boardGameRepository;

    @Mock
    private RateRepository rateRepository;

    @InjectMocks
    private BoardGameRatingService boardGameRatingService;

    @Test
    void when_game_is_rated_then_the_game_with_new_average_rate_is_returned() {
        when(boardGameRepository.findByName(POPULAR_GAME)).thenReturn(Optional.of(
                popularBoardGame()
        ));

        when(rateRepository.findByBoardGameName(POPULAR_GAME)).thenReturn(Arrays.asList(highRate(), lowRate()));

        final BoardGame newlyRatedGame = boardGameRatingService.ratingBoardGame(new RateBoardGame(POPULAR_GAME, lowRate().getRate()));

        assertEquals(new Rating(3.0d), newlyRatedGame.getRating());
    }


    @Test
    void when_requesting_games_with_higher_rate_then_3_only_higher_games_are_retrieved() {
        when(boardGameRepository.findAll()).thenReturn(Arrays.asList(
                popularBoardGame(),
                notPopularBoardGame()
        ));

        doReturn(Arrays.asList(highRate())).when(rateRepository).findByBoardGameName(POPULAR_GAME);
        doReturn(Arrays.asList(lowRate())).when(rateRepository).findByBoardGameName(NOT_POPULAR_GAME);

        final List<BoardGame> ratedHigherThan3 = boardGameRatingService.withHigherRateThan(3.0d);

        assertEquals(1, ratedHigherThan3.size());
        assertEquals(1, ratedHigherThan3.stream().filter(game -> game.getName().equals(POPULAR_GAME)).count());
    }

}

File: src/test/java/com/openvalue/boardgameratings/service/rating/BoardGameRatingServiceTest.kt
package com.openvalue.boardgameratings.service.rating

import com.openvalue.boardgameratings.api.Rating
import com.openvalue.boardgameratings.service.boardgame.BoardGameEntity
import com.openvalue.boardgameratings.service.boardgame.BoardGameRepository
import com.openvalue.boardgameratings.service.util.TestData.*
import junit.framework.Assert.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mockito.*
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.stubbing.OngoingStubbing
import org.mockito.stubbing.Stubber
import java.util.*
import kotlin.test.assertNotNull

@ExtendWith(MockitoExtension::class)
internal class BoardGameRatingServiceTest {

    private val boardGameRepository = mock(BoardGameRepository::class.java)

    private val rateRepository = mock(RateRepository::class.java)

    private val boardGameRatingService = BoardGameRatingService(boardGameRepository, rateRepository)

    @Test
    fun `when game is rated then the game with new average rate is returned`() {
        `when`(boardGameRepository.findByName(POPULAR_GAME))
                .returnPopularGame()

        `when`(rateRepository.findByBoardGameName(POPULAR_GAME))
                .returnHighAndLowRate()

        val (_, _, _, _, rating) = boardGameRatingService.ratingBoardGame(RateBoardGame(POPULAR_GAME, lowRate().rate))

        assertEquals(Rating(3.0), rating)
    }

    @Test
    fun `when requesting games with higher rate then 3 only higher games are retrieved`() {
        `when`(boardGameRepository.findAll()).thenReturn(listOf(
                popularBoardGame(),
                notPopularBoardGame()
        ))

        doReturn(listOf(highRate()))
                .whenPopularGameIsRetrieved()
        doReturn(listOf(lowRate()))
                .whenNotPopularGameIsRetrieved()

        val ratedHigherThan3 = boardGameRatingService.withHigherRateThan(3.0)

        assertEquals(1, ratedHigherThan3.size)
        assertNotNull(ratedHigherThan3.find { (name) -> name == POPULAR_GAME })
    }

    fun Stubber.whenPopularGameIsRetrieved() = this.`when`(rateRepository).findByBoardGameName(POPULAR_GAME)
    fun Stubber.whenNotPopularGameIsRetrieved() = this.`when`(rateRepository).findByBoardGameName(NOT_POPULAR_GAME)
    fun OngoingStubbing<Optional<BoardGameEntity>>.returnPopularGame() = this.thenReturn(Optional.of(popularBoardGame()))
    fun OngoingStubbing<List<RateEntity>>.returnHighAndLowRate() = this.thenReturn(listOf(highRate(), lowRate()))

}